### PR TITLE
[3/n] DataLoader2 initial support for randomness control

### PR DESCRIPTION
**Pull request by ejguan**
Fixes https://github.com/pytorch/data/issues/885

Add the support for DataLoader2 to control randomness over the pipeline:
- Implement SeedGenerator
  - `spawn` to generate sub-SeedGenerators for distributed workers
  - `generate_seed` to generate unique seeds
  - `generate_shared_seed` to generate distributed shared seeds
- Change API of ReadingService to take seed generator from DataLoader2. Then, the SeedGenerator of DataLoader2 becomes the source of truth of randomness within the whole data pipeline.

A separate PR will be added for online doc regarding determinism.

Differential Revision: D38947827

Last step for #885

### CODE REVIEW

**Comment by ejguan** [comment]
I might need to re-create a new PR via `ghexport` to support a stack of Diffs.

**Review by NivekT — APPROVED** [review]

File: torchdata/dataloader2/random/_philox.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import List, Optional, Tuple
+
+# Note [Philox Engine implementation]
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# Refer to: http://www.thesalmons.org/john/random123/papers/random123sc11.pdf for details regarding the engine.
+# Using Philox4×32-10 for the sake of performance, randomness and crush-resistance.
+# The following code could be optimized into C++ bindings
```
[the remark this answers is not in the document]
Comment (ejguan) [reply torchdata/dataloader2/random/_philox.py:13]: Yeah like one of my previous PR https://github.com/pytorch/pytorch/pull/62573, but we probably want to do it directly from TorchData side

File: torchdata/dataloader2/random/seed_generator.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+    def generate(self) -> int:
+        r"""
+        Generate one uint64 random seed
+        """
+        return self._rng.generate()
+
+    def spawn(self, process_id: int) -> "SeedGenerator":
+        r"""
+        Generate one uint64 random seed based on the provided process_id
```
[the remark this answers is not in the document]
Comment (ejguan) [reply torchdata/dataloader2/random/seed_generator.py:36]: Yeah. New ones will be sent to each worker process based on the worker id.

File: torchdata/dataloader2/reading_service.py [hunk 3]
```diff
@@ -301,26 +312,31 @@ def initialize(self, datapipe: DataPipe) -> DataPipe:
...
         Shares the same seed from rank 0 to other ranks across the distributed processes
         and apply the random seed to the graph of ``DataPipe``.
         """
-        # TODO: Seed Generator should be moved to DataLoader2 after the API
-        #       change of initialize_iteration is landed.
-        seed = self._share_seed()
-        _seed_generator = torch.Generator()
-        _seed_generator.manual_seed(seed)
+        shared_seed = self._share_seed(seed_generator.generate())
+        seed_generator.seed(shared_seed)
```
[the remark this answers is not in the document]
Comment (ejguan) [reply torchdata/dataloader2/reading_service.py:321]: We will broadcast the generated `shared_seed` on Rank 0 to other ranks. And, then we use this synchronized seed to re-seed DataLoader2 across all distributed processes.

File: torchdata/dataloader2/reading_service.py [hunk 4]
```diff
@@ -301,26 +312,31 @@ def initialize(self, datapipe: DataPipe) -> DataPipe:
...
-    def _share_seed(self):
-        shared_seed = torch.empty((), dtype=torch.int64).random_()
+    def _share_seed(self, seed: int):
+        # Convert uint64 to int64 to prevent overflow for integer Tensor
+        seed -= _HALF_UINT64
+        shared_seed = torch.tensor(seed, dtype=torch.int64)
         dist.broadcast(shared_seed, src=0, group=self._pg)
-        return shared_seed.item()
+        # Revert int64 back to uint64
+        return shared_seed.item() + _HALF_UINT64
```
[the remark this answers is not in the document]
Comment (ejguan) [reply torchdata/dataloader2/reading_service.py:339]: `shared_seed` will have the same value then adding `_HALF_UINT64` back to it shouldn't cause any difference

File: torchdata/dataloader2/dataloader2.py [hunk 5]
```diff
@@ -107,11 +108,18 @@ def __init__(
...
+        self._seed_generator = SeedGenerator()
+
     def __iter__(self) -> Iterator[T_co]:
         if self._terminated:
             raise Exception("Cannot iterate over the DataLoader as it has already been shut down")
 
         if self._reset_iter:
+
+            # TODO(ejguan): Provide an exclusive API to seed DataLoader
+            seed = int(torch.empty((), dtype=torch.int64).random_().item())
```
[the remark this answers is not in the document]
Comment (ejguan) [reply torchdata/dataloader2/dataloader2.py:120]: Yea.

File: torchdata/dataloader2/random/seed_generator.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (NivekT) [reply torchdata/dataloader2/random/seed_generator.py:36]: Makes sense. nit: Can you update the docstring?

[the remark this answers is not in the document]
Comment (ejguan) [reply torchdata/dataloader2/random/seed_generator.py:36]: Aha, Good catch!

File: torchdata/dataloader2/reading_service.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (NivekT) [reply torchdata/dataloader2/reading_service.py:321]: I see. nit: Should we rename `rng` to `dp_graph_rng` here to specify that it is specific for the DataPipe graph? It would be useful to distinguish `seed_generator` (which is for DL2) from `rng` (which is for the DP graph within DL2).

[the remark this answers is not in the document]
Comment (ejguan) [reply torchdata/dataloader2/reading_service.py:321]: I can rename it for now. And, I do have another follow-up PRs to change the behavior of `apply_random_seed` that would take `seed_generator`. I will write more detail reason in that PR.

**Review by NivekT — APPROVED** [review]

**Review by NivekT — APPROVED** [review]
